File: laboneq/compiler/seqc/prng_tracker.py
# Copyright 2022 Zurich Instruments AG
# SPDX-License-Identifier: Apache-2.0

from __future__ import annotations

from laboneq.compiler.seqc.seqc_generator import SeqCGenerator


class PRNGTracker:
    def __init__(self, seqc_gen: SeqCGenerator):
        self.seqc_gen: SeqCGenerator = seqc_gen
        self._range: int | None = None
        self._seed: int | None = None
        self._offset: int = 0
        self._committed: bool = False
        self._active_sample: str | None = None

    @property
    def range(self):
        return self._range

    @range.setter
    def range(self, value):
        assert not self._committed
        self._range = value

    @property
    def seed(self):
        return self._seed

    @seed.setter
    def seed(self, value):
        assert not self._committed
        self._seed = value

    @property
    def offset(self):
        return self._offset

    @offset.setter
    def offset(self, value: int):
        assert not self._committed
        self._offset = value

    @property
    def active_sample(self):
        return self._active_sample

    @active_sample.setter
    def active_sample(self, value: str):
        assert isinstance(value, str)
        assert self._active_sample is None, "must first drop existing sample"
        self._active_sample = value

    def drop_sample(self):
        assert self._active_sample is not None, "no sample to drop"
        self._active_sample = None

    def is_committed(self) -> bool:
        return self._committed

    def commit(self):
        assert not self._committed

        if self._seed is not None:
            self.seqc_gen.add_function_call_statement(
                name="setPRNGSeed", args=[self._seed]
            )
        if self._range is not None:
            self.seqc_gen.add_function_call_statement(
                name="setPRNGRange",
                args=[self._offset, self._offset + self._range - 1],
            )

        # the tracker now has been spent, so clear it to prevent another call to `commit()`
        self._committed = True

File: src/rust/seqc-tracker/src/prng_tracker.rs
// Copyright 2025 Zurich Instruments AG
// SPDX-License-Identifier: Apache-2.0

use crate::{seqc_generator::SeqCGenerator, seqc_statements::SeqCVariant};

#[derive(Default)]
pub struct PRNGTracker {
    range: Option<u32>,
    seed: Option<u32>,
    offset: u32,
    committed: bool,
}

impl PRNGTracker {
    pub fn new() -> Self {
        PRNGTracker {
            range: None,
            seed: None,
            offset: 0,
            committed: false,
        }
    }

    pub fn set_range(&mut self, value: u32) {
        assert!(!self.committed);
        self.range = Some(value);
    }

    pub fn set_seed(&mut self, value: u32) {
        assert!(!self.committed);
        self.seed = Some(value);
    }

    pub fn offset(&self) -> u32 {
        self.offset
    }

    pub fn set_offset(&mut self, value: u32) {
        assert!(!self.committed);
        self.offset = value;
    }

    pub fn is_committed(&self) -> bool {
        self.committed
    }

    pub fn commit(&mut self, seqc_gen: &mut SeqCGenerator) {
        assert!(!self.committed);
        if let Some(seed) = self.seed {
            seqc_gen.add_function_call_statement(
                "setPRNGSeed",
                vec![SeqCVariant::Integer(seed as i64)],
                None::<&str>,
            );
        }
        if let Some(range) = self.range {
            seqc_gen.add_function_call_statement(
                "setPRNGRange",
                vec![
                    SeqCVariant::Integer(self.offset as i64),
                    SeqCVariant::Integer((self.offset + range - 1) as i64),
                ],
                None::<&str>,
            );
        }
        // the tracker now has been spent, so clear it to prevent another call to `commit()`
        self.committed = true;
    }
}
